R/Plot_synergy.R

```diff
@@ -1,328 +1,327 @@
+      col.conc = .NumericApplyRange(colnames(scores.dose), col.range),
+      drug.row = paste(data$drug.pairs$drug.row[index], unit.text, sep = " "),
+      drug.col = paste(data$drug.pairs$drug.col[index], unit.text, sep = " "),
+      start.point = if(is.null(legend.start)) -color.range else legend.start,
+      end.point = if(is.null(legend.end)) color.range else legend.end)))
+}
 
-  }
+# Returns the highest absolute value in 'scores', rounded up to next ten.
+.RoundUpToNextTen <- function(scores) {
+  return(round(5 + max(abs(max(scores)),
+                       abs(min(scores))),
+               -1))
+}
 
-  if(!save.file) {
-    for(i in num.pairs) {
-      dev.new(noRStudioGD = TRUE)
-      replayPlot(plots[[i]])
-    }
+
+
+# Returns a string that may be used as a plot title.  The string includes
+# the mean value of all scores in the selected ranges, and the name of the
+# method used to compute them.  The first row and column are always removed, on
+# these assumptions:
+#  1. For each drug, the first recorded effect is for dose=0.
+#  2. When dose=0 for either drug (monotherapy), the synergy score is also 0.
+.ComputePlotTitle <- function(scores.dose, row.range, col.range, data.method) {
+  return(paste("Average synergy: ",
+               round(mean(scores.dose[.SliceAtLeastFirst(row.range),
+                                      .SliceAtLeastFirst(col.range)],
+                          na.rm = TRUE),
+                     3),
+               " (",data.method,")",
+               sep = ""))
+}
+
+# Convert 'val' to a slice.  However, guarantee that the slice starts at
+# 2 or higher.  So, both for val=c(1,5) and val=c(2,5), the function should
+# return 2:5.  The first two elements of list 'val' are taken to be the slice
+# endpoints; any other elements are ignored.  If 'val' is NULL, then the
+# returned slice is '-1' -- 'remove the first element'.
+.SliceAtLeastFirst <- function(val) {
+  return(if(is.null(val)) -1 else max(val[1],2):val[2])
+}
+
+# Returns a slice of 'val' specified by 'range', as numeric.
+# If 'range' is null, 'val' is wholly returned.  The first two elements of
+# 'range' are taken to be slice endpoints; any other values are ignored.
+# The desired portion of 'val' is converted to numeric form, then returned.  An
+# error occurs if any element of the selected slice of 'val' cannot be converted
+# to a number.
+.NumericApplyRange <- function(val, range) {
+  return(as.numeric(if(is.null(range)) val else val[range[1]:range[2]]))
+}
+
+# Returns a 2D slice of 'data' specified by 'row.range' and 'col.range'.  If
+# either range is NULL, the data for the corresponding axis of 'data' is
+# returned in full.  The row.range and col.range inputs are each treated as
+# lists whose first two elements are range endpoints, and whose remaining
+# elements (if any) are ignored.
+.ApplyRanges <- function(data, row.range, col.range) {
+  .RangeWithDefault <- function(range, default) {
```

Comment:
Just wondering if it's a good practice to write a function inside another function?

Reply:
There is nothing wrong with using an "inner function" generally, for languages that support that.  Typically, an inner function has two properties:

1. Its implementation can make use of variables in the scope that contains it.
2. It is only visible within the scope that contains it.

This one doesn't make particular use of point (1), so it's easily re-expressed at the top level.  So, I've done that.

Notice that I define *anonymous* inner functions in PlotSynergy() at top.  Those are necessary for the use of lapply(), and they *do* make use of point (1) as well as point (2).